quinn-udp/src/unix.rs

```diff
@@ -804,11 +854,18 @@ mod gso {
     // TODO: Add this to libc
     const UDP_SEGMENT: libc::c_int = 103;
 
+    // Global mutex to avoid calling `libc::uname` for each socket.
+    static SUPPORTED_BY_OS: Mutex<Option<bool>> = Mutex::new(None);
```

Comment:
Perhaps a `OnceLock` is more convenient here?

Reply:
We can then get rid of the `KernelVersion` struct and have a stateless function that checks if the version is < 4.18

Reply:
> We can then get rid of the KernelVersion struct and have a stateless function that checks if the version is < 4.18

I had the opposite plan to expand the `KernelVersion` struct by further spliting the parsing code into `KernelVersion::from_str` which would be called from `KernelVersion::from_uname` and write small tests with sting samples. The test could be a way of tracking if some kernels have inconsistent version strings as you noted here https://github.com/quinn-rs/quinn/issues/2246#issuecomment-2887779815.

This wouldn't be possible with a single function returning a boolean.

Let me know which way you prefer.

Reply:
> This wouldn't be possible with a single function returning a boolean.

You can split it into multiple functions and test the pure ones separately:

- 1 function that calls `libc::uname` and returns a string
- 1 function that checks if a given string is a kernel version > 4.18
- 1 function that combines the above too or perhaps just do that inline in the closure passed to `once_cell`

The 2nd one can then be easily unit-tested with multiple inputs.

I don't have a strong preference but if we can avoid intermediary structs, I think it is easier to understand.

Reply:
> I don't think we can use OnceLock for MSRV reasons, 

Running [`cargo msrv find`](https://github.com/foresterre/cargo-msrv) I get the result 1.74.1 (log: [msrv-find.log](https://github.com/user-attachments/files/20279627/msrv-find.log)). `OnceLock` seems to be [supported since 1.70.0](https://doc.rust-lang.org/std/sync/struct.OnceLock.html). So unless there is a plan to decrease the current MSRV I think `OnceLock` should be OK to use.

Reply:
Ah, okay, good.

Reply:
I did modify the code a bit, though I left the `KernelVersion` struct there. The reason was that I felt that the tests would be more to the point if they tested the parsing alone instead of parsing and checking >=4.18. It might be a matter of taste/style, if so, then perhaps it'd be better to leave that part to you (the code owners)?